token_examples/factory_token/src/token.rs

```diff
@@ -0,0 +1,423 @@
+    #[error("error in token: {0}")]
+    Token(#[from] TokenError),
+    /// Error from the underlying universal receiver hook library
+    #[error("error calling receiver hook: {0}")]
+    Receiver(#[from] ReceiverHookError),
+    /// Error from serialising data to RawBytes
+    #[error("ipld encoding error: {0}")]
+    Encoding(#[from] fvm_ipld_encoding::Error),
+    #[error("ipld blockstore error: {0}")]
+    Blockstore(#[from] ErrorNumber),
+    #[error("actor state not found {0}")]
+    NoState(#[from] NoStateError),
+    // deserialisation error when loading state
+    #[error("error loading state {0}")]
+    Deserialization(String),
+    // serialisation error when saving state
+    #[error("error saving state {0}")]
+    Serialization(String),
+    #[error("underlying state error {0}")]
+    State(#[from] StateError),
+    #[error("actor messaging error {0}")]
+    Messaging(#[from] MessagingError),
+    #[error("address not authorized")]
+    AddressNotAuthorized,
+    #[error("minting has been permanently disabled")]
+    MintingDisabled,
+}
+
+pub fn caller_address() -> Address {
+    let caller = sdk::message::caller();
+    Address::new_id(caller)
+}
+
+// NOTE: this is mostly lifted from the basic token actor we use in integration testing
+// differences are how state load/store is handled
+// and the build-in invoke() designed to be called from an upstream invoke()
+
+// this token implementation is designed to be embedded into some larger actor
+// where the state contains more than just the token
+#[derive(Serialize_tuple, Deserialize_tuple, Debug)]
+pub struct BasicToken {
+    /// Default token helper impl
+    pub token: TokenState,
+    /// basic token identifier stuff, should it go here or store separately alongside the state
+    pub name: String,
+    pub symbol: String,
+    pub granularity: u64,
+    /// address of authorised minting operator
+    pub minter: Option<Address>,
+}
+
+/// Implementation of the token API in a FVM actor
+///
+/// Here the Ipld parameter structs are marshalled and passed to the underlying library functions
+impl FRC46Token<RuntimeError> for BasicToken {
+    fn name(&self) -> String {
+        //String::from("FRC-0046 Token")
+        self.name.clone()
+    }
+
+    fn symbol(&self) -> String {
+        //String::from("FRC46")
+        self.symbol.clone()
+    }
+
+    fn granularity(&self) -> GranularityReturn {
+        //1
+        self.granularity
+    }
+
+    fn total_supply(&self) -> TotalSupplyReturn {
+        // TODO: token() wants mutable ref, we don't have one here and can't change the interface
+        // so need an immutable wrapper to call these things?
+        // or just bypass it and go to the state directly
+        //self.token().total_supply()
+        self.token.supply.clone()
+    }
+
+    fn balance_of(&self, params: Address) -> Result<BalanceReturn, RuntimeError> {
+        // TODO: same situation as total_supply
+        //Ok(self.token().balance_of(&params)?)
```

Comment:
Probably need a separate read-only implementation of the `Token` handle to clean this up (also for `total_supply` and any future query/getter functions)

Reply:
Addressed in https://github.com/helix-onchain/filecoin/pull/149#discussion_r1028709951 - I've changed the underlying trait to take mutable refs for these functions